text/3706-option-result-todo.md

```diff
@@ -0,0 +1,233 @@
+- Feature Name: `option-result-todo`
+- Start Date: 2024-10-03
+- RFC PR: [rust-lang/rfcs#3706](https://github.com/rust-lang/rfcs/pull/3706)
+- Rust Issue: [rust-lang/rust#3706](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes `Result::todo` and `Option::todo` functions which work like `.unwrap()` but imply different semantic reasons for the unwrapping.
+
+`.todo()` implies that error handling still needs to be done and that unwrapping here is a temporary thing that should be fixed in the future. This is analogous to `todo!()`.
+
+In short, this allows users to preserve semantic information about why they are unwrapping. This proves extremely useful in prototype code, and in distinguishing "todo, add error handling here" from "no, i actually want to panic in this case.".
+
+As an example:
+
+```rs
+// unwrap is still used for the analogous-to-panic!() case
+TcpListener::bind(&addr).unwrap();
+
+// we're panicking because error handling is not implemented yet.
+// this use case is common in prototype applications.
+let int: i32 = input.parse().todo();
+let arg2 = std::env::args().nth(2).todo();
+```
+
+# Motivation
+[motivation]: #motivation
+
+`.unwrap()` is semantically overloaded in rust. It finds itself used for two significantly different reasons:
+
+- If this is `None/Err`, our program is in a bad state and we want to exit (missing config files, missing resource files, some external invariant not upheld)
```

Comment:
I disagree about this reason. I think that `unwrap` does not mean to appear in production code outside tests -- we should always write `.expect("1 is non-zero")` (unreachable semantic) or `.expect("we are screwed because unable to get current exe")` (unrecoverable semantic)[^1]. 

So to me `.unwrap()` is already a `.todo()`: I just want to blindly unwrap it for now, I'm too lazy to give a reason, and maybe I'll check/replace them later. There is already an opt-in clippy lint [`clippy::unwrap-used`](https://rust-lang.github.io/rust-clippy/master/index.html#/unwrap_used) which does better than "highlight TODO in IDE".

[^1]: We could split these two semantics via something like `.expect_unreachable(...)`, but that's out of scope for the latest version of this RFC.

Reply:
then how about deprecating.unwrap() in a future edition and introduce .todo()

Reply:
> then how about deprecating.unwrap() in a future edition

That's not a good idea given its tremendous usage :thinking: "unwrap" is also more neutral than "todo" in case of somebody really want to stick on it (maybe for a reason, maybe not).